### ISSUE TITLE
Enchanted item deserialization issue

### ISSUE THREAD

**Issue by Shmeeb** [SpongePowered/SpongeForge#2208]
**I am currently running**

- SpongeForge version: 1.12.2-2705-7.1.0-BETA-3150
- Forge version: 14.23.4.2703
- Java version: 8
- Operating System: Ubuntu

Serialize an ItemStackSnapshot (that has enchantments) with:
`String json = DataFormats.JSON.write(snapshot.toContainer());`

Then load it with:
`DataContainer container = DataFormats.JSON.read(json);`
`ItemStackSnapshot snapshot = Sponge.getDataManager().deserialize(ItemStackSnapshot.class, container).get();`

The original and the deserialized version are not the same, as it seems a short turns into an int:
https://cdn.discordapp.com/attachments/452901983515443200/454297812243447809/unknown.png

You can use this code to test if the ItemStackSnapshot is identical according to Inventory#contains: https://pastebin.com/wKFbWzD8

**Comment by gabizou** [comment]
<img width="1565" alt="screen shot 2018-06-13 at 9 07 19 am" src="https://user-images.githubusercontent.com/1203877/41363617-361949ac-6ee9-11e8-8bbf-dc02d9caf2aa.png">
Running the latest SpongeForge (granted, in development) and I'm unable to reproduce the issue.

**Comment by dualspiral** [comment]
@gabizou 

The issue is that the enchantment level is a `short` in the original, when serialised and then deserialised, it is turned into an `int` - the type information is lost when converted into JSON and so when the data is read, there is no indication that it was meant to be a short.

When running `/cssave`, I breakpointed and checked the container.
![short-ench](https://user-images.githubusercontent.com/1904167/41370283-1e2bee70-6f3f-11e8-872e-d7f635da96aa.png)

Notice that `lvl` is a short.

I then run `/cscheck` and check the container.
![int-ench](https://user-images.githubusercontent.com/1904167/41370343-4aa4db38-6f3f-11e8-9028-fc56ac78e3ec.png)

Notice that `lvl` is now an integer.

I admit, I knew about this a while ago and have a hack in Nucleus, but then I completely forgot about it. It's because the JSON or HOCON doesn't contain the type of number that has been serialised, sure, serialising a short is easy, but the deserialiser doesn't know that the number is meant to be a short. Indeed, I had to do this because it causes item stacking issues. I thought it used to be that Minecraft had some issues with it being integers too, though I just tested giving an item from the JSON and it worked fine. You can see the stacking issues if you apply the enchantment to an item that stacks normally, though I haven't added that to the test plugin in this case.

Code I used is a modified version of Shmeeb's: https://gist.github.com/dualspiral/e52fc286d88cd9d5b80278b91c61eaa3 - `/csgive` to get the item.

I'm honestly not sure what we should do to solve this. My only thought is that we should attach some type data to the key or the value, but that's just going to get messy.

**Comment by Shmeeb** [comment]
Not sure if it helps but ItemStackComparators.ALL.compare seems to consider both snapshots as identical

**Comment by gabizou** [comment]
To be honest, I'm confused, because the test plugin I committed is succeeding normally without a hitch. It's matching the item stack to the item stack in the player's inventory, and it's deserializing just fine. Since `DataContainer` uses `Coerce` for all numbers, we can somewhat expect at least some understanding that deserializing a `DataSerializable` will still succeed, even if JSON converted a `short` to an `int`.

In short, I'm not sure what the actual issue with this is anymore. Sure, JSON is not storing the value type to deserialize without losing type information, but as long as type information is guaranteed by the implementation of the builder for the `DataSerializable`, it shouldn't really matter.
